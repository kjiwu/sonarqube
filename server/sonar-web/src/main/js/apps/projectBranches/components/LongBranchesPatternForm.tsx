/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Modal from 'react-modal';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import { setSimpleSettingValue, resetSettingValue, SettingValue } from '../../../api/settings';
import { LONG_BRANCH_PATTERN } from './LongBranchesPattern';

interface Props {
  onChange: () => void;
  onClose: () => void;
  project: string;
  setting: SettingValue;
}

interface State {
  submitting: boolean;
  value: string;
}

export default class LongBranchesPatternForm extends React.PureComponent<Props, State> {
  mounted: boolean;

  constructor(props: Props) {
    super(props);
    this.state = { submitting: false, value: props.setting.value || '' };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleSubmit = (event: React.SyntheticEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.setState({ submitting: true });
    const { value } = this.state;
    setSimpleSettingValue({
      key: LONG_BRANCH_PATTERN,
      value,
      component: this.props.project
    }).then(this.props.onChange, () => {
      if (this.mounted) {
        this.setState({ submitting: false });
      }
    });
  };

  handleValueChange = (event: React.SyntheticEvent<HTMLInputElement>) => {
    this.setState({ value: event.currentTarget.value });
  };

  handleCancelClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    this.props.onClose();
  };

  handleResetClick = (event: React.SyntheticEvent<HTMLButtonElement>) => {
    event.preventDefault();
    event.currentTarget.blur();
    this.setState({ submitting: true });
    resetSettingValue(LONG_BRANCH_PATTERN, this.props.project).then(this.props.onChange, () => {
      if (this.mounted) {
        this.setState({ submitting: false });
      }
    });
  };

  render() {
    const { setting } = this.props;
    const header = translate('branches.detection_of_long_living_branches');
    const submitDisabled = this.state.submitting || this.state.value === setting.value;

    return (
      <Modal
        isOpen={true}
        contentLabel={header}
        className="modal"
        overlayClassName="modal-overlay"
        onRequestClose={this.props.onClose}>
        <header className="modal-head">
          <h2>{header}</h2>
        </header>
        <form onSubmit={this.handleSubmit}>
          <div className="modal-body">
            <div className="big-spacer-bottom">
              {translate('branches.detection_of_long_living_branches.description')}
            </div>
            <div className="big-spacer-bottom">
              <input
                autoFocus={true}
                className="input-super-large"
                onChange={this.handleValueChange}
                required={true}
                type="text"
                value={this.state.value}
              />
              {setting.inherited && (
                <div className="note spacer-top">{translate('settings._default')}</div>
              )}
              {!setting.inherited &&
              setting.parentValue && (
                <div className="note spacer-top">
                  {translateWithParameters('settings.default_x', setting.parentValue)}
                </div>
              )}
            </div>
          </div>
          <footer className="modal-foot">
            {!setting.inherited &&
            setting.parentValue && (
              <button className="pull-left" onClick={this.handleResetClick} type="reset">
                {translate('reset_to_default')}
              </button>
            )}
            {this.state.submitting && <i className="spinner spacer-right" />}
            <button disabled={submitDisabled} type="submit">
              {translate('save')}
            </button>
            <a href="#" onClick={this.handleCancelClick}>
              {translate('cancel')}
            </a>
          </footer>
        </form>
      </Modal>
    );
  }
}
