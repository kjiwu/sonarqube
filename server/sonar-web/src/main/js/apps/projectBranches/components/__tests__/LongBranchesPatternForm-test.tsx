/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
jest.mock('../../../../api/settings', () => ({
  setSimpleSettingValue: jest.fn(() => Promise.resolve()),
  resetSettingValue: jest.fn(() => Promise.resolve())
}));

import * as React from 'react';
import { shallow } from 'enzyme';
import LongBranchesPatternForm from '../LongBranchesPatternForm';
import { change, submit, click } from '../../../../helpers/testUtils';

const setSimpleSettingValue = require('../../../../api/settings')
  .setSimpleSettingValue as jest.Mock<any>;

const resetSettingValue = require('../../../../api/settings').resetSettingValue as jest.Mock<any>;

beforeEach(() => {
  setSimpleSettingValue.mockClear();
  resetSettingValue.mockClear();
});

it('changes value', async () => {
  const onChange = jest.fn();
  const wrapper = shallow(
    <LongBranchesPatternForm
      onChange={onChange}
      onClose={jest.fn()}
      project="project"
      setting={{ inherited: true, key: '', value: 'release-.*' }}
    />
  );
  expect(wrapper).toMatchSnapshot();

  change(wrapper.find('input'), 'branch-.*');
  submit(wrapper.find('form'));
  expect(setSimpleSettingValue).toBeCalledWith({
    component: 'project',
    key: 'sonar.branch.longLivedBranches.regex',
    value: 'branch-.*'
  });

  await new Promise(setImmediate);
  expect(onChange).toBeCalled();
});

it('resets value', async () => {
  const onChange = jest.fn();
  const wrapper = shallow(
    <LongBranchesPatternForm
      onChange={onChange}
      onClose={jest.fn()}
      project="project"
      setting={{ inherited: false, key: '', parentValue: 'branch-.*', value: 'release-.*' }}
    />
  );
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('button[type="reset"]'));
  expect(resetSettingValue).toBeCalledWith('sonar.branch.longLivedBranches.regex', 'project');

  await new Promise(setImmediate);
  expect(onChange).toBeCalled();
});
