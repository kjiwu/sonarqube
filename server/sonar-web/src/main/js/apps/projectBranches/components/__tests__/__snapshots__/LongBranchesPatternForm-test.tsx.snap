// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes value 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="branches.detection_of_long_living_branches"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      branches.detection_of_long_living_branches
    </h2>
  </header>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-body"
    >
      <div
        className="big-spacer-bottom"
      >
        branches.detection_of_long_living_branches.description
      </div>
      <div
        className="big-spacer-bottom"
      >
        <input
          autoFocus={true}
          className="input-super-large"
          onChange={[Function]}
          required={true}
          type="text"
          value="release-.*"
        />
        <div
          className="note spacer-top"
        >
          settings._default
        </div>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <button
        disabled={true}
        type="submit"
      >
        save
      </button>
      <a
        href="#"
        onClick={[Function]}
      >
        cancel
      </a>
    </footer>
  </form>
</Modal>
`;

exports[`resets value 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="branches.detection_of_long_living_branches"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      branches.detection_of_long_living_branches
    </h2>
  </header>
  <form
    onSubmit={[Function]}
  >
    <div
      className="modal-body"
    >
      <div
        className="big-spacer-bottom"
      >
        branches.detection_of_long_living_branches.description
      </div>
      <div
        className="big-spacer-bottom"
      >
        <input
          autoFocus={true}
          className="input-super-large"
          onChange={[Function]}
          required={true}
          type="text"
          value="release-.*"
        />
        <div
          className="note spacer-top"
        >
          settings.default_x.branch-.*
        </div>
      </div>
    </div>
    <footer
      className="modal-foot"
    >
      <button
        className="pull-left"
        onClick={[Function]}
        type="reset"
      >
        reset_to_default
      </button>
      <button
        disabled={true}
        type="submit"
      >
        save
      </button>
      <a
        href="#"
        onClick={[Function]}
      >
        cancel
      </a>
    </footer>
  </form>
</Modal>
`;
